Initial pass at rendering golf=* features

jgruca opened the pull request:
Initial pass at rendering golf=* features. Based largely on [imagico/osm-carto-alternative-colors](https://github.com/imagico/osm-carto-alternative-colors/blob/alternative-colors/golf.mss).

Fixes gravitystorm/openstreetmap-carto#661

Changes proposed in this pull request:
- Update `landcover` layer to include golf area features
- Add `golf-lines` layer
- Add style/golf.mss and specify rendering for:
  - `golf=green` and `golf=tee` (light green color between `@grass` and `@park`;
    tee with point labeled with `ref`)
  - `golf=fairway` and `golf=driving_range` (`@grass`)
  - `golf=bunker` (color and pattern for `@sand` and beach)
  - `golf=rough` (darker green between `@scrub` and `@grass`)
  - `golf=hole` as way (dashed line, darker green derived from `@golf_course`; labeled with `ref`)
  - `golf=hole` as point and `golf=pin` (symbols/golf_pin.svg; labeled with `ref`)

Corrections, feedback, critiques welcome.

Examples:

https://www.openstreetmap.org/#map=15/41.7849/-93.7247

Before/After

Z15
![carto-1-zoom15](https://user-images.githubusercontent.com/1396120/94508486-b0c18100-01d7-11eb-9a59-322881219582.png)

Z16
![carto-1-zoom16](https://user-images.githubusercontent.com/1396120/94508531-c5057e00-01d7-11eb-9e01-44be911eb3b7.png)

Z17
![carto-1-zoom17](https://user-images.githubusercontent.com/1396120/94508551-ca62c880-01d7-11eb-9efb-02d55de57e6b.png)

https://www.openstreetmap.org/#map=16/41.8705/-93.2144

Before/After

Z14
![carto-2-zoom14](https://user-images.githubusercontent.com/1396120/94508586-e1a1b600-01d7-11eb-921d-8d7d56e6b9e1.png)

Z15
![carto-2-zoom15](https://user-images.githubusercontent.com/1396120/94508595-e6666a00-01d7-11eb-8d1e-f85c3ab316eb.png)

Z16
![carto-2-zoom16](https://user-images.githubusercontent.com/1396120/94508605-ea928780-01d7-11eb-8b8f-1b3fe85e48c6.png)

Z17
![carto-2-zoom17](https://user-images.githubusercontent.com/1396120/94508611-efefd200-01d7-11eb-82b0-ac4a8623bb56.png)

https://www.openstreetmap.org/#map=16/41.2767/-95.7367

Before/After

Z15
![carto-3-zoom15](https://user-images.githubusercontent.com/1396120/94509958-2e3ac080-01db-11eb-8b9a-a26fae4585d9.png)

Z16
![carto-3-zoom16](https://user-images.githubusercontent.com/1396120/94509964-31ce4780-01db-11eb-97ec-cbdfb9ff3d05.png)

Z17
![carto-3-zoom17](https://user-images.githubusercontent.com/1396120/94509968-35fa6500-01db-11eb-98f6-56d43fc323a5.png)

jeisenbe reviewed (changes requested):
Thanks for working on this, @jgruca 
I think it will be beneficial to show these features, and this initial draft already looks pretty good.
But there are a few lines in the code that can be simpified, and I would like us to think about the color choices. It would be best to avoid adding too many new fill colors for areas, and any new colors chosen need to be intuitive and not easily confused with other features. This can be tricky to get just right.

imagico commented:
Thanks for taking on working on this rather complicated topic.

As @jeisenbe mentioned the choice of colors is not going to work - we already have a real lot of green polygon fill colors in this style so introducing even more of them is not a good idea.  The green line color is also going to be confusing with the color of bridleways.  I would suggest trying to do this without introducing new colors.

For reference - the color scheme used in the ac-style can be seen [here](https://wiki.openstreetmap.org/w/index.php?title=Tag:natural%3Dreef&diff=2040640):

* the overall golf course color is unified with the campsite color (removing one more green tone from this style)
* fairway and driving_range are in grass color
* green is in pitch color
* bunker is using sand color
* rough is using a fine grain pattern mixing grass and scrub
* symbols and lines are in black.

Line labels for artificial features in `@oblique-fonts` is also not matching our normal text styling traditions.

Note there are quite a few subtle implications in the code from the ac-style regarding interpretation of the tagging that would need to be discussed - in particular what variants of tagging ref are supported and how to render them.  For consistency with the rest of the style you should probably also not use a compound geometry type layer but unify points and polygons using ST_PointOnSurface() and separate out the line features to a different layer.

jgruca commented on style/golf.mss:
```diff
@@ -0,0 +1,66 @@
+  ::high-zoom[zoom >= 12] {
+    [feature = 'golf_green'],
+    [feature = 'golf_tee'] {
+      polygon-fill: mix(@grass, @park, 50%);
```
Makes sense. I have replaced the `mix()` with a `lighten()` on `@grass`, which I feel better represents the golf green. If you would prefer a specific color definition instead of a modification, I can do that too.

jgruca commented on style/golf.mss:
```diff
@@ -0,0 +1,66 @@
+      polygon-pattern-file: url('symbols/beach.png');
+    }
+    [feature = 'golf_rough'] {
+      polygon-fill: mix(@scrub, @grass, 50%);
```
I have replaced this `mix()` with `@grass`, overlaid with the `golf_rough.svg` pattern from the ac-style. The result is a subtle differentiation rather than a large color difference.

jgruca commented:
I have pushed up some changes to try to address some of the comments above.

* SVG files have been cleaned up and specific color references moved to .mss file
* Reduced the number of new colors introduced to just one, a lightened `@grass` for the green
* Updated the golf course color to the same as campsite
* tee, fairway, rough, and driving_range are now all `@grass` color
* The `golf=hole` line is now a solid `@leisure-green` line, labeled with a book font
* Added a texture for `golf=rough`

I have no particular attachment to these colors and I am happy to adjust as needed, as they still differ a bit from the ac-style you linked above.

If you could point me to examples/docs of separating line features on a separate layer, or further elaborate on that, I will certainly give that a try.

Here are updated screenshots, same locations as above.
![carto-1-zoom15](https://user-images.githubusercontent.com/1396120/94628252-2df6ff80-0285-11eb-96c9-269453b4b123.png)
![carto-1-zoom16](https://user-images.githubusercontent.com/1396120/94628253-2e8f9600-0285-11eb-8171-4a783c9dfe60.png)
![carto-1-zoom17](https://user-images.githubusercontent.com/1396120/94628256-2e8f9600-0285-11eb-8179-b6093fcdddf4.png)

![carto-2-zoom15](https://user-images.githubusercontent.com/1396120/94628257-2f282c80-0285-11eb-900e-47deaafe85d9.png)
![carto-2-zoom16](https://user-images.githubusercontent.com/1396120/94628259-2f282c80-0285-11eb-966d-2a36fe71e6c4.png)
![carto-2-zoom17](https://user-images.githubusercontent.com/1396120/94628260-2f282c80-0285-11eb-84d5-7ecdd6cc1d0e.png)

![carto-3-zoom15](https://user-images.githubusercontent.com/1396120/94628262-2fc0c300-0285-11eb-8197-24720150db7b.png)
![carto-3-zoom16](https://user-images.githubusercontent.com/1396120/94628264-2fc0c300-0285-11eb-8c88-27129960c593.png)
![carto-3-zoom17](https://user-images.githubusercontent.com/1396120/94628266-2fc0c300-0285-11eb-886c-75f0006c249e.png)

imagico commented:
> If you could point me to examples/docs of separating line features on a separate layer, or further elaborate on that, I will certainly give that a try.

The golf-lines layer combines point, polygon and linestring features in a single layer.  This we so far don't do anywhere else in this style.  The common practice is to combine points and polygons for label/symbol rendering using ST_PointOnSurface() - see for example https://github.com/gravitystorm/openstreetmap-carto/blob/b10aef3866bacf387581b8fea4eec265010b0d14/project.mml#L1316 and for linestring features using a separate layer - see amenity-points:

https://github.com/gravitystorm/openstreetmap-carto/blob/b10aef3866bacf387581b8fea4eec265010b0d14/project.mml#L1444

and amenity-line:

https://github.com/gravitystorm/openstreetmap-carto/blob/b10aef3866bacf387581b8fea4eec265010b0d14/project.mml#L1656

You should also consider integrating point symbol/label rendering into the normal point symbol layer because otherwise you will get difficulties with correct prioritization.

jgruca commented:
> The golf-lines layer combines point, polygon and linestring features in a single layer. This we so far don't do anywhere else in this style. The common practice is to combine points and polygons for label/symbol rendering using ST_PointOnSurface()

So in this case, `golf=hole`, `pin`, and `tee` points should be moved into a separate layer, leaving `golf=hole` linestrings in the golf-lines layer. I imagine this can drastically simplify the SQL for golf-lines. Should the `ref` label for `golf=hole` be moved into the text-line layer?

> You should also consider integrating point symbol/label rendering into the normal point symbol layer because otherwise you will get difficulties with correct prioritization.

Is the amenity-points layer considered the normal point symbol layer here?

Organizationally, does it still make sense to have a separate .mss file just for golf features, or would you prefer to have style rules moved into the files more closely associated with their layers?

jgruca commented:
I have attempted to address some of the issues with these latest changes.
* `golf-line`, as a layer, now contains just `golf=hole` as a line
* `golf=hole` and `golf=pin`, as points, have been moved to `amenity-points`. Please let me know if this is not the appropriate layer.
* The rendering for the `ref` label of `golf=hole` has been moved to the `text-line` layer
* The point previously rendered in the center of `golf=tee` polygons has been removed entirely, as I haven't found any other golf course maps that show this information
* The rendering for `golf=rough` has been tweaked a little by setting opacity on the pattern; see the screenshot below.

![carto](https://user-images.githubusercontent.com/1396120/95037548-0ba02000-0691-11eb-8042-87e358d96dbc.png)

I have made changes to the SQL for some of these layers with limited knowledge of the size or indexes of these tables, so if there are optimizations or changes that are needed there please let me know.

tpetillon commented:
Wouldn't be more consistent to use `@stadium` (or `@leisure`, as they're identical) for the background? Semantically, a golf course is closer to a football pitch or a tennis court than a camp site.

It would look like this:
![image](https://user-images.githubusercontent.com/204831/95511319-98135100-09b7-11eb-82d2-ac2fb68554e0.png)
(Copying the outline of other sports facilities as well.)

Going further, `@leisure` and `@campsite` could probably be merged, decreasing the number of shades of green in the style.

jgruca commented:
I have no issues with a switch to `@leisure` for the background color. How do others feel?

imagico commented:
Sorry, but no, this does not work for me at all.  The introduction of the color that is defined as `@leisure` as well as its extension as a residual color for a broad range of semantically largely unrelated features - playground, fitness_station, recreation_ground (but not park!), dog_park, stadium, water_park - was done without consensus and countered previous efforts in designing a more consistent, systematic and homogeneous color scheme.  See #2268, #2891, #2954,  #2964, #3843, https://github.com/gravitystorm/openstreetmap-carto/issues/2250#issuecomment-240053481, #2270.

If you want to have a common color for different kinds of sports infrastructure that is fine with me - provided that the color suggested for that fits with the overall color system.  But engrossing playground, dog_park and recreation_ground with that will not work.

Regarding your modified color scheme - i think your choice of color for green is too close to the campsite color:

http://davidjohnstone.net/pages/lch-lab-colour-gradient-picker#def6c0,daf0c4

Using the pitch color seems the natural choice here - and in my eyes it also would highlight the structure of a golf course.  For rough - you would need to check how it relates to the other darker greens in our style, in particular scrub and wood.  A rough on a golf course does not feature taller vegetation, it is usually just longer grass so if the color is closer to scrub or wood than to grass then it is possibly quite misleading.

Generally speaking given the problems with adding more colors to the style for general readability i would expect you to support any introduction of a new color or color-pattern combination to be supported with convincing reasoning.  We cannot support this merely as a local convenience choice.  For rough the need for distinction between the cut grass of a fairway and the longer grass for the rough area is a clear necessity.  For bunker and green i see no convincing arguments so far.

jgruca commented:
Thanks for the direction there. I was unaware of the history of some of the color choices, and appreciate the context. I have reverted from `@leisure` back to `@campsite`.

I'd like this PR to limit itself to rendering `golf=*` features rather than a comprehensive refactor of sports infrastructure.

My hesitation in using `@pitch` as the color for the golf green is that 1) a sports pitch normally includes the full area of play and 2) on a course with water hazards, it is a blue enough green that it might be confused for water:

![image](https://user-images.githubusercontent.com/1396120/95923208-a330fd00-0d7a-11eb-98e3-dd915279d7c9.png)

Having said that, you are correct in saying that the current `golf=green` color is too close to `@campground`. In fact, I propose that simply switching the color to use `@campground` accomplishes the twin goals of differentiating the green from the fairway and decreasing the number of colors in the style.

I infer from your comments a pattern for `golf=bunker` is unnecessary. I have removed that.

Since a number of things have changed since the first commit, I am summarizing the style changes in this PR below.

| Tag | Example | Summary of changes | Justification |
| --- | --- | --- | --- |
| `leisure=golf_course`<br>`leisure=miniature_golf` | ![image](https://user-images.githubusercontent.com/1396120/95932320-2a896b00-0d91-11eb-9c9d-9e4a6afe0f7b.png) | Fill now `@campground` | Removing `@golf_course`, a single-use color. Semantically, a campground could be considered similar to a golf course: both are large areas, usually featuring natural vegetation, paths, and maintained areas for recreational activities. |
| `golf=green` | ![image](https://user-images.githubusercontent.com/1396120/95932362-48ef6680-0d91-11eb-8da3-60094bceec89.png) | Fill now `@campground`, outline now `@grass` | A golf green is an important feature to differentiate from the fairway. Using `@campground` allows contrast against the `@grass` of the fairway and does not introduce a new color. Providing an outline guarantees it will stand out even when it is wholly separate from the fairway. |
| `golf=tee`<br>`golf=fairway`<br>`golf=driving_range` | ![image](https://user-images.githubusercontent.com/1396120/95932426-72a88d80-0d91-11eb-942d-7edc64f120bb.png) ![image](https://user-images.githubusercontent.com/1396120/95932346-3b39e100-0d91-11eb-914a-76935a55ac07.png) | Fill now `@grass` | Filling these areas with `@grass` -- their standard surface -- formalizes rendering of these tags, which are frequently mistagged as `landuse=grass` so they will render at all. |
| `golf=bunker` | ![image](https://user-images.githubusercontent.com/1396120/95932301-1e051280-0d91-11eb-87b0-0c9ec4188f55.png) | Fill now `@sand` | Filling these areas with `@sand` matches `natural=sand` and will help discourage those who would mistag bunkers as `natural=sand` for rendering purposes. |
| `golf=rough` | ![image](https://user-images.githubusercontent.com/1396120/95932401-63c1db00-0d91-11eb-94b8-1c2e0435909a.png) | Fill now `@grass`, pattern now `golf_rough.svg` | Using a pattern in these areas differentiates between the textures of two kinds of grass on the course without introducing a new color. |
| `golf=hole` (way) | ![image](https://user-images.githubusercontent.com/1396120/95932631-fbbfc480-0d91-11eb-9b98-bec3429792e4.png) | Line with stroke of `@leisure-green`, labeled with `ref` of hole | Labels the hole and direction of play; the color matches golf course label. |
| `golf=hole` (node)<br>`golf=pin` | ![image](https://user-images.githubusercontent.com/1396120/95932382-54429200-0d91-11eb-80d2-a1a75c4c1832.png) | Icon `golf_pin.svg`, labeled with `ref` of hole, both `@leisure-green` color | Labels the actual hole, called out with an icon for visibility, in a color matching other new labeling. |

I think the only potential disagreements here are 1) the color of `golf=green` and 2) the current pattern on `golf=rough` along with its effect on the perceived color of that area. The changes I have outlined here are merely personal preferences, and if you tell me that changing them to x and y will result in a merged PR, I will make those changes.

imagico commented:
Thanks for the additional details on your reasoning.  I get your point about the pitch color's limited contrast to water - which is however not limited to use on golf courses.  Cause of this is the pitch color used here (see #2363) - the ac-style uses a different color which does not have this problem.  Regarding the argument that leisure=pitch usually covers the full area a sport is played on - that is not universal, table tennis is an obvious example, long distance running where most of it happens on the street while the finish is on a stadium track is another.

The issue with using the same color for golf=green and leisure=golf_course is that there would be no visible distinction between golf=green being mapped and not being mapped.

The use for the same color for campsite and golf_course is in my eyes not so much a case of functional similarity but of classical overloading.  Both are rendered with a symbol in addition to the polygon fill so there is very little risk of confusion.  This is not ideal but a reasonable solution IMO given the color constraints.

jgruca commented:
> The issue with using the same color for golf=green and leisure=golf_course is that there would be no visible distinction between golf=green being mapped and not being mapped.

Yes, this is a valid concern. I have addressed it in the current PR by specifying an outline (`line-width: 2; line-color: @grass;`) around the green. In my admittedly-unscientific survey of local golf courses, there were three configurations of fairway and green that I saw. Most common was when the green was totally surrounded by fairway; in this case, there is no difference in rendering. Second most common was when the fairway was adjacent to and sharing nodes with the green. In this case, the outline blends well with the fairway and serves to delineate the opposite side of the green. And in the case you describe, where a green is placed totally separate from the fairway, the green is outlined with the grass color to distinguish it from the surrounding course.

![image](https://user-images.githubusercontent.com/1396120/96068953-cc2cbd00-0e62-11eb-8c01-b49b5ed9b55d.png) ![image](https://user-images.githubusercontent.com/1396120/96069451-e0bd8500-0e63-11eb-8317-130e2dff1905.png)
![image](https://user-images.githubusercontent.com/1396120/96069045-04cc9680-0e63-11eb-9ff6-bc4640c7f844.png) ![image](https://user-images.githubusercontent.com/1396120/96069434-d7341d00-0e63-11eb-9b7f-4cdf950716df.png)
![image](https://user-images.githubusercontent.com/1396120/96069301-87edec80-0e63-11eb-9fed-29bd71dccda2.png) ![image](https://user-images.githubusercontent.com/1396120/96069414-cdaab500-0e63-11eb-884a-a86f0f180c3e.png)

imagico commented:
Though visually this idea looks smooth it suffers from several problems IMO:

* it does not highlight the significance of the green for the game.
* the line signature can be easily mistaken for a mapped geometry (a slim grass/fairway strip around the green) which impairs intuitive mapper feedback.
* it still does not provide functional feedback on the green neing properly mapped in case the green is fully surrounded by a fairway polygon.

jgruca commented:
How would you feel about `@pitch` for `tee`, `fairway`, and `driving_range`, and `@grass` for `green`?

![image](https://user-images.githubusercontent.com/1396120/96201257-bf22d300-0f21-11eb-9b44-2bb9d0bde02d.png)

I'm still not a huge fan of `@pitch`, however, this accomplishes:

* The areas of the map colored `@pitch` are less likely to be the same size/shape as smaller bodies of water
* The larger area covered in that color helps drive home that the areas are all associated with one another, especially if `golf=hole` is mapped on them
* `golf=green` is visible apart from the fairway and distinguishable from a cutout in a fairway

You could still make the argument that then `golf=green` is not distinguishable from a simple grass area, and that using `@grass` confers no sports-related semantics. However, this is the same situation you have when `golf=fairway` is `@grass` and `golf=green` is `@pitch`, just reversed.

jgruca commented:
I should note in this scenario, `golf=rough` would also be `@pitch`, but with the same texture as before.

![image](https://user-images.githubusercontent.com/1396120/96202185-2f325880-0f24-11eb-8c79-dc7f405f89c0.png)

imagico commented:
I tried using the pitch color for the fairways in the ac-style and i did not very much - but it was a different color of course.  Generally speaking design wise and independent of semantics it looks usually better if you use stronger colors for the smaller features and less strong colors for the larger features.  But in terms of color systematics your examples seem to make some sense since the pitch color then indeed indicates the area the sport is physically played on.

Other opinions would be welcome.

Adamant36 commented:
> Other opinions would be welcome.

I had fooled around with rendering golf features a while back and getting the colors right was a problem. I think the colors you and jgruca have came up with look good though. Its at least a lot better then what I tried.

jeisenbe commented:
Semantically, I don’t think it makes sense to use the pitch color plus a texture to represent areas of `golf=rough` while using `@grass` for `golf=green` While I can imagine a fairway as being closely related to a pitch, the rough is not the main playing surface. From what I can tell, it seems to mean “areas of longer grass, which are not part of the fairway, green or tee, but are still inside of the out-of-bounds markers”, so I’m not convinced that this is clearly different from other areas like `landuse=grass` or `landuse=meadow`.

jgruca commented:
I think the combination of green, fairway, rough, and tee areas are analogous to a sports pitch: they comprise the playable, in-bounds area of the sport. The goal here is to differentiate between those same areas, preferably without using new colors.

For `golf=green`: certainly `@grass` is not the optimal choice, but perhaps it is the best choice of the available greens currently within the style. Do you think a slightly-lightened `@pitch` would be considered? I am currently operating under the assumption that introducing new colors will further complicate this PR's approval process.

For `golf=rough`: I don't believe any existing sub-pitch features are rendered for other sports, and certainly not using textures. However, I feel like this is an appropriate use of a texture: the main difference between the rough and the fairway is literally its texture; the rendered output is subtle yet indicative of the difference; no new colors need to be introduced.

imagico commented:
@jeisenbe - yes, that is a valid point.  One of the arguments for using pitch color for the green and grass color for the fairway as i did in the ac-style is that this is what mappers quite frequently tag for the renderer - 40 percent of all golf=green features have a leisure=pitch tag and >90 percent of all golf=fairway have a landuse=grass tag.  This does not prelude us rendering things differently but it indicates that mappers are content with this rendering or with the semantic implication of the green being kind of like a pitch and the fairway being something like a generic grass area (or with both).

Another thought:  Rendering details of a golf course is kind of at the edge of what is sensible within this style considering its goals.  The functional aspects of the detailed golf course tagging are of interest only for a small fraction of the map users, a large part of why it might none the less make sense to render this here is that golf courses are large features and appear kind of empty without details so having these shown kind of helps balancing the map even for those who are not interested in the sport.  For that purpose a rendering oriented at least roughly at physical appearance aspects and not exclusively on functional things could be advisable

gmar5 commented:
Hello. I intrude in the discussion to offer the perspective of somebody who has no interest in golf. I got curious about these issues after noticing the heavy misuse of tags like `landuse=grass` and `natural=sand`.

I personally think that rendering of overly-fine details within golf courses seriously risks being detrimental to legibility and clarity for users of the map who are not looking at those areas.

Important goals in my view are:
- the golf course area should remain readable as a coherent single unit;
- it should be easily distinguishable from nearby areas, such as natural features or parks, both in urban and rural contexts.

I thought that the current unique shade of green did that moderately well. Unless it is entirely covered by parcels of grass, wood, sand, etc., at which point nothing of distinctive is left.

If that colour needs to be changed, I think it should converge, based on function, with other sport and recreation areas, and not be confused with nature-related ones. (I also don't think that disapproval of the current `@recreation_ground` green should be a deterrent from the idea of aligning them in principle, but I am not privy to what seemed already lengthy discussions on that, so I'll leave this be.)
The campsite green in itself is not bad, since it is quite unique and the two are distinguished by icons. Also using the semantically sport-related `@pitch` green for all the larger play areas while minimizing the use of `@grass` green seems a wise direction.

Rendering a lot of the micro-level components will inevitably reduce the distinctive clarity of the golf course, both because a large area is filled up and broken up, becoming visually busier, and because of more chance of colours being shared with nearby areas outside. Therefore, I think they should not be overdone, but if they are valued by users who care about golf, then perhaps should they be rendered only at closer zoom levels? 

This is not a criticism of the recent proposals above, which could actually work, I just wanted to remind you not to work on them in isolation, but to think about (and simulate) this within the broader map contexts of cities and countryside where golf courses sit, from the point of view of the more general user.

I hope this perspective can be somewhat helpful. Thanks for your work.

jeisenbe commented:
@gmar re: "Rendering a lot of the micro-level components will inevitably reduce the distinctive clarity of the golf course”

I also have no personal interest in golf, but I think this concern is misplaced. 

The shapes of fairways and other features on a golf course is rather distinctive and quicklly recognizable even for those of us who have never played the game. Since there are many shades of green in use in this style, legibility will actually be improved by rendering the sand traps, greens and fairways, since the semi-standardized shapes and positions of these objects will make it clear that “this area is a golf course” to map users with even a passing familiarity with the concept.

jgruca commented:
It has been a few months now. Are there any other opinions, objections, or suggestions regarding this PR? Anything I can do to move this forward?

jeisenbe commented:
@jgruca I still think it does not work to use `pitch` color plus a texture to represent areas of `golf=rough` while using `@grass` for `golf=green`, and @imagico has also mentioned some arguements in favor of using `@grass` for the fairway and `@pitch` for the green ( https://github.com/gravitystorm/openstreetmap-carto/pull/4212#issuecomment-719384811 and https://github.com/gravitystorm/openstreetmap-carto/pull/4212#issuecomment-710226909).

jgruca commented:
I see. So is there consensus for using `@grass` for `golf=fairway/tee/driving_range` and `@pitch` for `golf=green`? Personally, I still prefer the idea of using `@pitch` for the fairway (see my reasoning https://github.com/gravitystorm/openstreetmap-carto/pull/4212#issuecomment-709671171) which @imagico mentioned made some sense (https://github.com/gravitystorm/openstreetmap-carto/pull/4212#issuecomment-710226909). However, I'm willing to swap those colors if the concerns aren't warranted.

That just leaves the matter of how to handle `golf=rough`. @jeisenbe, you specifically mention `@pitch` with a texture doesn't work for you; would `@grass` with a texture be acceptable?

![rough-texture](https://user-images.githubusercontent.com/1396120/107998976-eb239300-6fab-11eb-8c4f-6867a07b98ed.png)

jgruca commented:
I have updated the PR with the previously-mentioned proposal; namely, `@grass` for `golf=fairway/tee/driving_range/rough` and `@pitch` for `golf=green`. Additionally, `golf=rough` has a texture applied to differentiate it from the fairway.

pnorman commented:
With the update you're now including unrelated commits in your branch. It looks like a screwed up rebase happened.

jgruca commented:
Oops, yes, this includes me rebasing my changes on top of master. Given the number of changes that have occurred in this PR, would it be easier to open a new PR with the current squashed set of changes and an accurate summary at the top?

pnorman commented:
I think that'd be easiest.
